### ISSUE TITLE
mpremote: all mpremote commands are followed by a hard reboot

### ISSUE THREAD

**Issue by Josverl** [micropython/micropython#9659]
according to @jimmo  mpremote should be able to resume a repl session on a connected mcu without a reboot by using the following command pattern:
``` sh
mpremote connect exec "start_something()"
mpremote connect resume exec "continue_here()"
```
where the `resume` is the important part.
see [Discord discussion](https://discord.com/channels/574275045187125269/1029416691488264252/1031558675900407818) 

however when I try this : 
 1. I see that any variables set in the the first `mpremote` commands are lost 
 2. I see that the esp32 performs a soft reset after exiting the `raw REPL` 
     (observed through logic analyzer)

So even though the 2nd mpremote command does not force a reset, by that time a soft reboot has already occurred.
I also see a `soft reset` after: 

- exiting the repl
- eval "42"
- exec "print(42)"

## Actual result on windows : 
```
C:\develop\usb>mpremote exec "foo=bar=42;print(foo)"
42

C:\develop\usb>mpremote resume exec "print(bar)"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'bar' isn't defined
```
## Expected 
the MCU not to reset between two mpremote commands similar to the behavour on linux 

```
$ mpr exec "foo=bar=42; print(foo)"
42
$ mpr exec "print(bar)"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'bar' isn't defined
$ mpr exec "foo=bar=42; print(foo)"
42
$ mpr resume exec "print(bar)"
42
```

results as seen by a logic analyzer hooked up to the RX/TX pins of the esp32 
![image](https://user-images.githubusercontent.com/981654/196273853-8dcf6a8d-9942-495d-ad7e-6234ef50400c.png)

``` text
\x01\x03
>>> 
>>> 
>>> 
>>> 
raw REPL; CTRL-B to exit
\x02OKx05A\x01foo=bar=43;print(foo)\x04
MPY: soft reboot
raw REPL; CTRL-B to exit
>R\x01\x80\0\x01\x0443
\x04\x04>ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x17 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4540
ho 0 tail 12 room 4
load:0x40078000,len:12448
load:0x40080400,len:4124
entry 0x40080680
MicroPython v1.19.1 on 2022-06-18; 4MB/OTA module with ESP32
Type "help()" for more information.
>>> 
>>> 
>>> 
>>> 
>>> 
raw REPL; CTRL-B to exit
\x02R\x01\x80\0\x01\x04\x04Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'foo' isn't defined
\x04>ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x17 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4540
ho 0 tail 12 room 4
load:0x40078000,len:12448
load:0x40080400,len:4124
entry 0x40080680
MicroPython v1.19.1 on 2022-06-18; 4MB/OTA module with ESP32
Type "help()" for more information.
>>> 
```
## other observations 
- same behavior is observed with multiple firmware versions, and on different boards 
- esp8266 devices do NOT soft-reset in the same way  ( note: all my esp8266 devices use a different serialport driver) 
- 

## Versions 

 - mpremote : 0.4.0 
 - pyserial: 3.5
 - Python 3.9.13
 - Version: 1.72.2 (user setup)
 - OS: Windows_NT x64 10.0.22621
 - Micropython - v1.19.1
 - MCU - all esp32 
 - serial port driver esp32 : Silicon Laboratories Inc. 11.1.0.53
 - serial port driver esp8266 : wch.cn - No version or other details 
```
> mpremote devs
COM9 01665F52 10c4:ea60 Silicon Labs None # esp32
COM11 0001 10c4:ea60 Silicon Labs None # esp32 
COM20 00000000A 0403:6001 FTDI None # UART debug 
COM21  1a86:7523 wch.cn None # ESP8266
```

### firmwares 
 - mpremote connect com9 get-info
{'ver': 'v1.19.1', 'build': '', 'platform': 'esp32', 'port': 'esp32', 'machine': 'ESP32 module (spiram) with ESP32', 'release': '1.19.1', 'nodename': 'esp32', 'name': 'micropython', 'family': 'micropython', 'sysname': 'esp32', 'version': '1.19.1'}
 - mpremote connect com11 get-info
   {'ver': 'v1.19.1', 'build': '', 'platform': 'esp32', 'port': 'esp32', 'machine': '4MB/OTA module with ESP32', 'release': '1.19.1', 'nodename': 'esp32', 'name': 'micropython', 'family': 'micropython', 'sysname': 'esp32', 'version': '1.19.1'}
 - mpremote connect com21 get-info
   {'ver': 'v1.19.1', 'build': '', 'platform': 'esp8266', 'port': 'esp8266', 'machine': 'ESP module (1M) with ESP8266', 'release': '1.19.1', 'nodename': 'esp8266', 'name': 'micropython', 'family': 'micropython', 'sysname': 'esp8266', 'version': '1.19.1'}

oldest firmware tested with same behavior:
 - mpremote connect com9 get-info
   {'ver': 'v1.10', 'build': '', 'platform': 'esp32', 'port': 'esp32', 'machine': 'ESP32 module with ESP32', 'release': '1.10.0', 'nodename': 'esp32', 'name': 'micropython', 'family': 'micropython', 'sysname': 'esp32', 'version': '1.10.0'}

**Comment by Josverl** [comment]
if you want to take a look at the trace :
[three_device_session.zip](https://github.com/micropython/micropython/files/9805167/three_device_session.zip)

open with : https://www.saleae.com/downloads/

**Comment by jimmo** [comment]
@Josverl This is a hard reset, not a soft reset. A soft reset would not print out the esp boot sequence.

The only hint I have is that on many ESP32 boards, they wire up RTS/CTS from the on-board cp210x / ftdi in a way such that esptool.py can trigger the board to reset into the bootloader. I wonder if somehow the way the serial port is being opened on Windows is also triggering this?

**Comment by Josverl** [comment]
you are correct , there is both a soft and a hard reset. I was blinded by the first one.
but im sure that  I and students in bootcamps have been able to connect to and disconnect many times to a board without it rebooting.

my esp32s all use a CP210x ,
while my esp8266 all have an CH340 

but that is about the most common UART/USB interface used, 

Could it be that the state of the DTR/RTS signals have an effect ?
One fix that I have needed to add to a [few terminal tools](https://github.com/pycom/pymakr-vsc/issues?q=is%3Aissue+is%3Aclosed+DTR) is to compensate for the fact that that windows and linux actually handle the DTR/RTS signals differently. (neither is wrong, just different) 
(I've not looked into how pyserial handles this yet)

**Comment by mgsb** [comment]
I encountered this recently with a new-to-me board, SparkFun MicroMod ESP32. The [schematic](https://cdn.sparkfun.com/assets/2/2/5/9/5/MicroMod_ESP32_Schematic.pdf) documents the DTR/RTS behavior for resetting:

* If DTR is LOW, toggling RTS from HIGH to LOW resets to run mode
* If RTS is HIGH, toggling DTR from LOW to HIGH resets to bootloader

As a test, I changed the Pyboard constructor in `pyboard.py` to assert the DTR and RTS pins:

``` diff
@@ -275,7 +275,11 @@ class Pyboard:
             delayed = False
             for attempt in range(wait + 1):
                 try:
-                    self.serial = serial.Serial(device, **serial_kwargs)
+                    self.serial = serial.Serial(**serial_kwargs)
+                    self.serial.dtr = True
+                    self.serial.rts = False
+                    self.serial.port = device
+                    self.serial.open()
                     break
                 except (OSError, IOError):  # Py2 and Py3 have different errors
                     if wait == 0:
```

This fixed the problem on Linux and Windows, but not macOS (Big Sur, Monterrey).

On Linux, using

``` shell
stty -F /dev/ttyUSB0 -hup
```

also prevented the hard reset.

**Comment by jimmo** [comment]
Thanks @mgsb -- From what I've read elsewhere this seems like the right approach, although frustrating about macOS.

I haven't had much chance to look into this in detail but one thing I'd like to understand -- is this a new issue, or has it always been a problem? i.e. was this always causing DTR/RTS to bounce quickly, but there was enough capacitance and/or pull strength that it was never noticed? Are we seeing it now because just a handful of boards have less capacitance on these lines?

@Josverl does this solve it for your board?

**Comment by cymplecy** [comment]
Following long debugging session on Discord today (22Dec2022) I can confirm I had this issue on an ESP32-C3FH4 (ESP32-C3 with RBG matrix on board) and that the fix above fixed my issue

Who do I see about getting it incorporated into standard code base? :)

**Comment by Josverl** [comment]
I'll give this a test, but the .py fix is like what I've been doing in .js for pymskr, but with the added condition/constraint to only manipulate dtr rts on win32/64

**Comment by jimmo** [comment]
> Following long debugging session on Discord today (22Dec2022) I can confirm I had this issue on an ESP32-C3FH4 (ESP32-C3 with RBG matrix on board) and that the fix above fixed my issue

Some more details -- this was Windows, and the behaviour was that there was a reset on _disconnect_.

The C3 is interesting because it's on-chip USB/UART, rather than external, but clearly it's internally wired up the same way with the DTR/RTS pins.

> Who do I see about getting it incorporated into standard code base? :)

I will send a PR. I think it sounds like this change doesn't solve everything, but overall only improves things and doesn't make anything break. But will need some more testing.

**Comment by andrewleech** [comment]
> I will send a PR. I think it sounds like this change doesn't solve everything, but overall only improves things and doesn't make anything break. But will need some more testing.

Lucky me found the case where it does break :-D on windows and stm32 with "custom" driver installed (with zadig) to give the ports a personalised name! So yeah not a common situation, though common to every device I make at work... 

I'm not sure how to handle it better here though really, other than adding a extra bit of extra injected code on connection to query `os.uname()` to enable stuff like this per-port. Of course this would fail if the device isn't at repl on connection (though much of mpremote fails in this circumstance).

In theory I might be able to update my driver creation / installation to keep the manufacturer as "Microsoft" to bypass this bit of mpremote code too, though that doesn't help devices / drivers already in the field...

**Comment by jimmo** [comment]
@andrewleech What actually broke? The issue here was that it triggered a hard reset of the board due to the ESP32 bootloader-activation-via-DTR/RTS-thing. I can't see how that would apply to an stm32 board?

**Comment by andrewleech** [comment]
For me, having `self.serial.dtr = False  # DTR False = gpio0 High = Normal boot` causes no data to be sent from STM to PC, so I get no console content come in. Text I type goes out, but is blocking as the STM is respecting the flow control and waiting for RTS to go high :-)

**Comment by crackwitz** [comment]
This problem has reappeared.

This is the current code:

https://github.com/micropython/micropython/blob/ebc9525c953dfbb6c44ab99320b5278e4314dafe/tools/mpremote/mpremote/transport_serial.py#L68-L82

And this is my fix that works for me, against an ESP32-C3 that appears as a COM port with a driver by *Microsoft*:

```python
if portinfo and portinfo[0].manufacturer != "Microsoft":
    # ESP8266/ESP32 boards use RTS/CTS for flashing and boot mode selection.
    # DTR False: to avoid using the reset button will hang the MCU in bootloader mode
    # RTS False: to prevent pulses on rts on serial.close() that would POWERON_RESET an ESPxx
    self.serial.dtr = False  # DTR False = gpio0 High = Normal boot
    self.serial.rts = False  # RTS False = EN High = MCU enabled
else: # my addition
    self.serial.dtr = True
    self.serial.rts = False
```

Please let me know if I should open a new issue or PR for this.

**Comment by dpgeorge** [comment]
@Josverl any comment on the above issue?

**Comment by Josverl** [comment]
Im away from home , but happen to have a ESP32-C3  AI0C3 board with me that has 2x USB-C.

I can reproduce the behavior on the USB-CDC port only, the issue does not occur on on the usb-port labeled UART.
that matched the current detection mechanism , that on purpose only sets the DTR/RTS for 'non Microsoft drivers'
That is used as a simple, ( too simple ... ) detection mechanism to distinguish ESP from other boards.

from my earlier research for https://github.com/micropython/micropython/issues/11075
the `DTR=True&RTS=False` does work for STM32,SAMD51 and RP2040, but the mpremote code is starting to feel a bit like a magic incantation
I'd prefer to have a way to better identify if a board is an ESPxxxx , and therefore needs non-default DTR/RTS settings to avoid interfering with other boards that use the same signals for other purposes. 
ideally from the Composite device descriptor as that does identify the vender and manufacturer as :

```
idVendor                 : 0x303A (Espressif Incorporated)
iManufacturer            : 0x01 (String Descriptor 1)
 Language 0x0409         : "Espressif°"
```

the simplest solution I can find is to use Espressif's VID to determine if DTR/RTS should be handled.

``` py
VID_ESPRESSIF = 0x303A  # Espressif Incorporated
....
                    portinfo = list(serial.tools.list_ports.grep(device))  # type: ignore
                    if portinfo and ( 
                        portinfo[0].vid == VID_ESPRESSIF or
                        portinfo[0].manufacturer != "Microsoft"
                        ):
                        # ESP8266/ESP32 boards use RTS/CTS for flashing and boot mode selection.
                        # DTR False: to avoid using the reset button will hang the MCU in bootloader mode

``` 

Ill send a PR for this and another bug in the errno handling I found while testing.

<details><summary>USB Composite Device</summary>
<p>


```


    =========================== USB Port3 ===========================

Connection Status        : 0x01 (Device is connected)
Port Chain               : 1-6-3
Properties               : 0x01
 IsUserConnectable       : yes
 PortIsDebugCapable      : no
 PortHasMultiCompanions  : no
 PortConnectorIsTypeC    : no
ConnectionIndex          : 3
CompanionIndex           : 0
 CompanionHubSymLnk      : USB#VID_045E&PID_09BA#MSFT30002315322806#{f18a0e88-c30c-11d0-8815-00a0c906bed8}
 CompanionPortNumber     : 3
 -> CompanionPortChain   : 2-3-3

      ======================== USB Device ========================

        +++++++++++++++++ Device Information ++++++++++++++++++
Device Description       : USB Composite Device
Device Path              : \\?\USB#VID_303A&PID_1001#34:85:18:02:8C:F8#{a5dcbf10-6530-11d2-901f-00c04fb951ed} (GUID_DEVINTERFACE_USB_DEVICE)
Kernel Name              : \Device\USBPDO-8
Device ID                : USB\VID_303A&PID_1001\34:85:18:02:8C:F8
Hardware IDs             : USB\VID_303A&PID_1001&REV_0101 USB\VID_303A&PID_1001
Driver KeyName           : {36fc9e60-c465-11cf-8056-444553540000}\0018 (GUID_DEVCLASS_USB)
Driver                   : \SystemRoot\System32\drivers\usbccgp.sys (Version: 10.0.26100.3912  Date: 2025-04-26)
Driver Inf               : C:\WINDOWS\inf\usb.inf
Legacy BusType           : PNPBus
Class                    : USB
Class GUID               : {36fc9e60-c465-11cf-8056-444553540000} (GUID_DEVCLASS_USB)
Service                  : usbccgp
Enumerator               : USB
Location Info            : Port_#0003.Hub_#0003
Location IDs             : PCIROOT(0)#PCI(1400)#USBROOT(0)#USB(6)#USB(3), ACPI(_SB_)#ACPI(PC00)#ACPI(XHCI)#ACPI(RHUB)#ACPI(HS06)#USB(3)
Container ID             : {bc780716-f39c-57ce-aaae-40eb6a7d1d25}
Manufacturer Info        : (Standard USB Host Controller)
Capabilities             : 0x94 (Removable, UniqueID, SurpriseRemovalOK)
Status                   : 0x0180600A (DN_DRIVER_LOADED, DN_STARTED, DN_DISABLEABLE, DN_REMOVABLE, DN_NT_ENUMERATOR, DN_NT_DRIVER)
Problem Code             : 0
Address                  : 3
HcDisableSelectiveSuspend: 0
EnableSelectiveSuspend   : 0
SelectiveSuspendEnabled  : 0
EnhancedPowerMgmtEnabled : 0
IdleInWorkingState       : 0
WakeFromSleepState       : 0
Power State              : D0 (supported: D0, D2, D3, wake from D0, wake from D2)
 Child Device 1          : USB JTAG/serial debug unit (WinUsb Device)
  Device ID              : USB\VID_303A&PID_1001&MI_02\6&B1CC194&0&0002
  Class                  : USBDevice
  Driver KeyName         : {88bae032-5a81-49f0-bc3d-a4ff138216d6}\0000 (GUID_DEVCLASS_USBDEVICE)
  Service                : WINUSB
 Child Device 2          : USB Serial Device (COM8)
  Device Path            : \\?\USB#VID_303A&PID_1001&MI_00#6&b1cc194&0&0000#{86e0d1e0-8089-11d0-9ce4-08003e301f73} (GUID_DEVINTERFACE_COMPORT)
  Kernel Name            : \Device\00000cd4
  Device ID              : USB\VID_303A&PID_1001&MI_00\6&B1CC194&0&0000
  Class                  : Ports
  Driver KeyName         : {4d36e978-e325-11ce-bfc1-08002be10318}\0007 (GUID_DEVCLASS_PORTS)
  Service                : usbser
  COM-Port               : COM8 (\Device\USBSER000)

        +++++++++++++++++ Registry USB Flags +++++++++++++++++
HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\usbflags\303A10010101
 osvc                    : REG_BINARY 01 10
 SkipContainerIdQuery    : REG_BINARY 01 00
 NewInterfaceUsage       : REG_DWORD 00000000 (0)

        ---------------- Connection Information ---------------
Connection Index         : 0x03 (3)
Connection Status        : 0x01 (DeviceConnected)
Current Config Value     : 0x01
Device Address           : 0x11 (17)
Is Hub                   : 0x00 (no)
Device Bus Speed         : 0x01 (Full-Speed)
Number Of Open Pipes     : 0x05 (5 pipes to data endpoints)
Pipe[0]                  : EndpointID=2  Direction=IN   ScheduleOffset=0  Type=Interrupt
Pipe[1]                  : EndpointID=2  Direction=OUT  ScheduleOffset=0  Type=Bulk
Pipe[2]                  : EndpointID=3  Direction=IN   ScheduleOffset=0  Type=Bulk
Pipe[3]                  : EndpointID=1  Direction=OUT  ScheduleOffset=0  Type=Bulk
Pipe[4]                  : EndpointID=1  Direction=IN   ScheduleOffset=0  Type=Bulk

        --------------- Connection Information V2 -------------
Connection Index         : 0x03 (3)
Length                   : 0x10 (16 bytes)
SupportedUsbProtocols    : 0x03
 Usb110                  : 1 (yes)
 Usb200                  : 1 (yes)
 Usb300                  : 0 (no)
 ReservedMBZ             : 0x00
Flags                    : 0x00
 DevIsOpAtSsOrHigher     : 0 (Is not operating at SuperSpeed or higher)
 DevIsSsCapOrHigher      : 0 (Is not SuperSpeed capable or higher)
 DevIsOpAtSsPlusOrHigher : 0 (Is not operating at SuperSpeedPlus or higher)
 DevIsSsPlusCapOrHigher  : 0 (Is not SuperSpeedPlus capable or higher)
 ReservedMBZ             : 0x00

    ---------------------- Device Descriptor ----------------------
bLength                  : 0x12 (18 bytes)
bDescriptorType          : 0x01 (Device Descriptor)
bcdUSB                   : 0x200 (USB Version 2.00) -> wrong, device is Full-Speed only
bDeviceClass             : 0xEF (Miscellaneous)
bDeviceSubClass          : 0x02
bDeviceProtocol          : 0x01 (IAD - Interface Association Descriptor)
bMaxPacketSize0          : 0x40 (64 bytes)
idVendor                 : 0x303A (Espressif Incorporated)
idProduct                : 0x1001
bcdDevice                : 0x0101
iManufacturer            : 0x01 (String Descriptor 1)
 Language 0x0409         : "Espressif°"
iProduct                 : 0x02 (String Descriptor 2)
 Language 0x0409         : "USB JTAG/serial debug unit°"
iSerialNumber            : 0x03 (String Descriptor 3)
 Language 0x0409         : "34:85:18:02:8C:F8"
bNumConfigurations       : 0x01 (1 Configuration)

    ------------------ Configuration Descriptor -------------------
bLength                  : 0x09 (9 bytes)
bDescriptorType          : 0x02 (Configuration Descriptor)
wTotalLength             : 0x0062 (98 bytes)
bNumInterfaces           : 0x03 (3 Interfaces)
bConfigurationValue      : 0x01 (Configuration 1)
iConfiguration           : 0x00 (No String Descriptor)
bmAttributes             : 0xC0
 D7: Reserved, set 1     : 0x01
 D6: Self Powered        : 0x01 (yes)
 D5: Remote Wakeup       : 0x00 (no)
 D4..0: Reserved, set 0  : 0x00
MaxPower                 : 0xFA (500 mA)

        ------------------- IAD Descriptor --------------------
bLength                  : 0x08 (8 bytes)
bDescriptorType          : 0x0B
bFirstInterface          : 0x00
bInterfaceCount          : 0x02
bFunctionClass           : 0x02 (Communications and CDC Control)
bFunctionSubClass        : 0x02
bFunctionProtocol        : 0x00
iFunction                : 0x00 (No String Descriptor)

        ---------------- Interface Descriptor -----------------
bLength                  : 0x09 (9 bytes)
bDescriptorType          : 0x04 (Interface Descriptor)
bInterfaceNumber         : 0x00
bAlternateSetting        : 0x00
bNumEndpoints            : 0x01 (1 Endpoint)
bInterfaceClass          : 0x02 (Communications and CDC Control)
bInterfaceSubClass       : 0x02 (Abstract Control Model)
bInterfaceProtocol       : 0x00 (No class specific protocol required)
iInterface               : 0x00 (No String Descriptor)

        -------------- CDC Interface Descriptor ---------------
bFunctionLength          : 0x05 (5 bytes)
bDescriptorType          : 0x24 (Interface)
bDescriptorSubType       : 0x00 (Header Functional Descriptor)
bcdCDC                   : 0x110 (CDC Version 1.10)

        -------------- CDC Interface Descriptor ---------------
bFunctionLength          : 0x04 (4 bytes)
bDescriptorType          : 0x24 (Interface)
bDescriptorSubType       : 0x02 (Abstract Control Management Functional Descriptor)
bmCapabilities           : 0x02
 D7..4                   : 0x00 (Reserved)
 D3                      : 0x00 (not supports the notification Network_Connection)
 D2                      : 0x00 (not supports the request Send_Break)
 D1                      : 0x01 (supports the request combination of Set_Line_Coding, Set_Control_Line_State, Get_Line_Coding, and the notification Serial_State)
 D0                      : 0x00 (not supports the request combination of Set_Comm_Feature, Clear_Comm_Feature, and Get_Comm_Feature)

        -------------- CDC Interface Descriptor ---------------
bFunctionLength          : 0x05 (5 bytes)
bDescriptorType          : 0x24 (Interface)
bDescriptorSubType       : 0x06 (Union Functional Descriptor)
bControlInterface        : 0x00
bSubordinateInterface[0] : 0x01

        -------------- CDC Interface Descriptor ---------------
bFunctionLength          : 0x05 (5 bytes)
bDescriptorType          : 0x24 (Interface)
bDescriptorSubType       : 0x01 (Call Management Functional Descriptor)
bmCapabilities           : 0x03
 D7..2                   : 0x00 (Reserved)
 D1                      : 0x01 (can send/receive call management information over a Data Class interface)
 D0                      : 0x01 (handles call management itself)
bDataInterface           : 0x01

        ----------------- Endpoint Descriptor -----------------
bLength                  : 0x07 (7 bytes)
bDescriptorType          : 0x05 (Endpoint Descriptor)
bEndpointAddress         : 0x82 (Direction=IN EndpointID=2)
bmAttributes             : 0x03 (TransferType=Interrupt)
wMaxPacketSize           : 0x0040 (64 bytes)
bInterval                : 0x01 (1 ms)

        ---------------- Interface Descriptor -----------------
bLength                  : 0x09 (9 bytes)
bDescriptorType          : 0x04 (Interface Descriptor)
bInterfaceNumber         : 0x01
bAlternateSetting        : 0x00
bNumEndpoints            : 0x02 (2 Endpoints)
bInterfaceClass          : 0x0A (CDC-Data)
bInterfaceSubClass       : 0x02
bInterfaceProtocol       : 0x00
iInterface               : 0x00 (No String Descriptor)

        ----------------- Endpoint Descriptor -----------------
bLength                  : 0x07 (7 bytes)
bDescriptorType          : 0x05 (Endpoint Descriptor)
bEndpointAddress         : 0x01 (Direction=OUT EndpointID=1)
bmAttributes             : 0x02 (TransferType=Bulk)
wMaxPacketSize           : 0x0040 (64 bytes)
bInterval                : 0x01 (ignored)

        ----------------- Endpoint Descriptor -----------------
bLength                  : 0x07 (7 bytes)
bDescriptorType          : 0x05 (Endpoint Descriptor)
bEndpointAddress         : 0x81 (Direction=IN EndpointID=1)
bmAttributes             : 0x02 (TransferType=Bulk)
wMaxPacketSize           : 0x0040 (64 bytes)
bInterval                : 0x01 (ignored)

        ---------------- Interface Descriptor -----------------
bLength                  : 0x09 (9 bytes)
bDescriptorType          : 0x04 (Interface Descriptor)
bInterfaceNumber         : 0x02
bAlternateSetting        : 0x00
bNumEndpoints            : 0x02 (2 Endpoints)
bInterfaceClass          : 0xFF (Vendor Specific)
bInterfaceSubClass       : 0xFF
bInterfaceProtocol       : 0x01
iInterface               : 0x00 (No String Descriptor)

        ----------------- Endpoint Descriptor -----------------
bLength                  : 0x07 (7 bytes)
bDescriptorType          : 0x05 (Endpoint Descriptor)
bEndpointAddress         : 0x02 (Direction=OUT EndpointID=2)
bmAttributes             : 0x02 (TransferType=Bulk)
wMaxPacketSize           : 0x0040 (64 bytes)
bInterval                : 0x01 (ignored)

        ----------------- Endpoint Descriptor -----------------
bLength                  : 0x07 (7 bytes)
bDescriptorType          : 0x05 (Endpoint Descriptor)
bEndpointAddress         : 0x83 (Direction=IN EndpointID=3)
bmAttributes             : 0x02 (TransferType=Bulk)
wMaxPacketSize           : 0x0040 (64 bytes)
bInterval                : 0x01 (ignored)

    ----------------- Device Qualifier Descriptor -----------------
Error                    : ERROR_GEN_FAILURE

      -------------------- String Descriptors -------------------
             ------ String Descriptor 0 ------
bLength                  : 0x04 (4 bytes)
bDescriptorType          : 0x03 (String Descriptor)
Language ID[0]           : 0x0409 (English - United States)
             ------ String Descriptor 1 ------
bLength                  : 0x16 (22 bytes)
bDescriptorType          : 0x03 (String Descriptor)
Language 0x0409          : "Espressif°"  *!*ERROR  contains 1 NULL character
             ------ String Descriptor 2 ------
bLength                  : 0x38 (56 bytes)
bDescriptorType          : 0x03 (String Descriptor)
Language 0x0409          : "USB JTAG/serial debug unit°"  *!*ERROR  contains 1 NULL character
             ------ String Descriptor 3 ------
bLength                  : 0x24 (36 bytes)
bDescriptorType          : 0x03 (String Descriptor)
Language 0x0409          : "34:85:18:02:8C:F8"
             -- MSFT String Descriptor 0xEE --
bLength                  : 0x12 (18 bytes)
bDescriptorType          : 0x03 (String Descriptor)
qwSignature Lang 0x0409  : "MSFT100"
bMS_VendorCode           : 0x10 (GET_MS_DESCRIPTOR=0x10)
bPad                     : 0x00


```


</p>
</details>

**Comment by crackwitz** [comment]
The device I'm dealing with is [an ESP32-C3 board from aliexpress](https://www.aliexpress.com/item/1005005967641936.html) with one USB-C port and it claims to be VID 303A, PID 1001. I did not try other DTR/RTS combinations since the first one I tried worked already.

**Comment by Josverl** [comment]
@andrewleech, 
did you find the root cause for your problem ?
if it was indeed the RTS held high then perhaps its also possible to prevent that if there is a common VID for STM devices ?

**Comment by Josverl** [comment]
> VID 303A, PID 1001. 

These VID:PID are same as for the ESP32-C3 board I have.

and for a ESP32-C2 - single Connector board: 

      idVendor                 : 0x303A (Espressif Incorporated)
      idProduct                : 0x4001
      bcdDevice                : 0x0100
      

So checking for the VID should work, unless board manufacturers replace it with their own VID's.
then we would need to add a config option / file.

**Comment by Josverl** [comment]
> I did not try other DTR/RTS combinations since the first one I tried worked already

@crackwitz, the challenge is not only fixing this for your board, 
but doing that without breaking mpremote + other boards across a number use different usecases.

**Comment by crackwitz** [comment]
I am aware of that, which is why I strive to make perfectly clear what I *didn't* try.

**Comment by andrewleech** [comment]
> @andrewleech, 
> did you find the root cause for your problem ?
> if it was indeed the RTS held high then perhaps its also possible to prevent that if there is a common VID for STM devices ? 

My particular issue was from having a custom vid pid and separately loaded driver that did not have Microsoft as the listed manufacturer so it didn't match the checks in code.

I tried making changes to this block of code to resolve the problems better but every option I tried had other subtle negative outcomes too, they're in a PR here somewhere I think.

**Comment by Josverl** [comment]
> did not have Microsoft as the listed manufacturer so it didn't match the checks in code.

Thanks for clarifying. The only way that we can avoid that is by having a list of  regexes to match the known driver names/vendors for the uart drivers. That is the approach we ended up using for Pymakr. Simpler than gathering all possible VIDs for the serial UART drivers. 

```py
[ 		"Pycom.*", 		"FTDI", 		"Microsoft", 		"Microchip Technology, Inc.", 		"Silicon Labs",  "Espressif.*", ....	]
```

I think that should be a separate PR though

**Comment by projectgus** [comment]
This is probably a silly question as I'm not across the whole thing, but in the limit is it reasonable to patch the opposite case? (i.e. don't clear DTR/RTS for drivers/configs where we know that will cause a problem, do it for all other boards which should include most/all ESP32 boards plus a bunch of boards which don't care either way.

Alternatively, if it's too hard to get all of the custom cases then maybe this needs to be an option on `mpremote resume` (i.e. `mpremote resume --esp ...` or similar.) - in addition to detecting the most obvious cases.

**Comment by Josverl** [comment]
not silly , but hard to parse your inverse logic, but I think we may see the same way ahead.

let me try to recap my experience, and intents

- pyboard/pyserial defaults are OK on linux / macOS, and were mostly broken on Windows. ( different world visions between the OS ecosystems) 
- this mostly surfaces as spikes in the DTR/CTSlines at either connection , but more often at disconnection 
- this is most problematic in boards that use these to do special things, such as the ESPxxx boards that connect these lines to some logic and !EN and other strapping Pins
- it is not always simple to see / trace these signals - but on the ESP32 / ESP8266 chips it often is, so I used there for my testing.
- most of these boards have UART-USB convertors on board - with multiple different drivers and driver versions for each of these. 
Naming of these drivers is - not quite consistent, and new drivers are released .

To avoid ESPxx board to reset on disconnect , or hang in bootloader on connect we need the set DTR/CTS to False/False **Only for these boards**, both for UART/USB as well as for USB/CDC , but there is no authoritative list of VIDs / vendor names to do so.

With Pycom I was able to iterate quickly on user feedback - and updates were done in weeks, so we were able to add regexes for vendor strings that were reported quickly.
with mpremote some of the same updates have been stalled for months, or longer. That is the main reason I have chosen for the if not "Microsoft" driver approach.

Now with the modern ESPxxx boards switching to native USB , and less reliance on 3rd party UART drivers I think it is a good time to switch tactics 

- The default mpremote config should have a list of VIDs-PIDs that require "ESP-Handling" on Windows
- this config should be extendible/overridable  though the mpremote configuration file (currently still broken on windows ) 
- when connecting to a port - mpremote can check if it needs ESP-Handling - and then act accordingly 

*  if / when needed new vids/pids can be added to the default list via PRs, and included in the next release 
*  personal / interim changes can be done by editing the mpremote config file.

**Comment by dpgeorge** [comment]
> * personal / interim changes can be done by editing the mpremote config file.

I think that would be good, to be able to customise/configure/override the default behaviour based on user configuration.

**Comment by Josverl** [comment]
Ok , I can create a draft PR for that, so we can iterate over the details.

then we can move the conversation there as well.
